matchzoo/models/esim.py

```diff
@@ -0,0 +1,218 @@
+
+        encoder_layer = keras.layers.Bidirectional(keras.layers.LSTM(
+            hidden_size,
+            return_sequences=True,
+            dropout=self._params['dropout_rate']
+        ))
+        # Encode the two text
+        # shape = [B, L, 2*H]
+        encoded_left = encoder_layer(embed_left)
+        # shape = [B, R, 2*H]
+        encoded_right = encoder_layer(embed_right)
+        # Make interaction and get the interactive representation
+        # shape = [B, L, 2*H]
+        # shape = [B, R, 2*H]
+        attended_left, attended_right = self._attention(encoded_left, left_mask, encoded_right, right_mask)
+        # shape = [B, L, 8*H]
+        enhanced_left = keras.layers.Concatenate()([encoded_left, attended_left,
+                                                    keras.layers.Subtract()([encoded_left, attended_left]),
+                                                    keras.layers.Multiply()([encoded_left, attended_left])])
+        # shape = [B, R, 8*H]
+        enhanced_right = keras.layers.Concatenate()([encoded_right, attended_right,
+                                                     keras.layers.Subtract()([encoded_right, attended_right]),
+                                                     keras.layers.Multiply()([encoded_right, attended_right])])
+        # Project to the H size
+        projection = keras.layers.Dense(hidden_size, activation=keras.layers.ReLU())
+        # shape = [B, L, H]
+        projected_left = projection(enhanced_left)
+        # shape = [B, R, H]
+        projected_right = projection(enhanced_right)
+        projected_left = keras.layers.Dropout(
+            rate=self._params['dropout_rate'])(projected_left)
+        projected_right = keras.layers.Dropout(
+            rate=self._params['dropout_rate'])(projected_right)
+        # Fusion the enhanced features by another BiLSTM
+        fusion_layer = keras.layers.Bidirectional(keras.layers.LSTM(
+            hidden_size,
+            return_sequences=True,
+            dropout=self._params['dropout_rate']
+        ))
+        # shape = [B, L, 2*H]
+        fusion_left = fusion_layer(projected_left)
+        # shape = [B, R, 2*H]
+        fusion_right = fusion_layer(projected_right)
+        # Do global avg and max pooling and get a fixed-length representation vector
+        avg_left_rep = keras.layers.Lambda(lambda x: K.sum(x * expand_tensor(left_mask), axis=1) /
+                                                     expand_tensor(K.sum(left_mask, axis=1)))(fusion_left)
+        avg_right_rep = keras.layers.Lambda(lambda x: K.sum(x * expand_tensor(right_mask), axis=1) /
+                                                      expand_tensor(K.sum(right_mask, axis=1)))(fusion_right)
+        max_left_rep = keras.layers.Lambda(lambda x: K.max(x + (1 - tile_tensor(left_mask)) * -1e7, axis=1))(
+            fusion_left)
+        max_right_rep = keras.layers.Lambda(lambda x: K.max(x + (1 - tile_tensor(right_mask)) * -1e7, axis=1))(

```

Comment:
can we set `-1e7` as an epsilon paramterer?

Reply:
This layer is  equal to masked 1d-MaxPooling，-1e7  option is to set the padding position's value to -inf.